src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRunCodegenNodeFactory.cs

```diff
@@ -30,6 +32,8 @@ public sealed class ReadyToRunCodegenNodeFactory : NodeFactory
             });
         }
 
+        public PEReader PEReader;
```

Comment:
Nit: Given how things tend to go, this field is going to lead to a design where we hardcode the fact there's only one input file into too many places because it's too convenient to access it from here.

Reply:
There are basically two classes like that - this one and SignatureContext (which will arrive with the node factory refactoring). My current thinking is that we should basically switch these over every time we compile a method because the method module provides the reference token resolution context unless I'm mistaken. I agree it's going to be a non-trivial refactoring but I'm worried this change is complex enough already.

Reply:
> we should basically switch these over every time we compile a method

Do we have a good spot for the switch over? I'm worried about multi-threaded compilation. The CoreRT compiler is not multithreaded yet, but we kept that in mind with all the design choices Project X compiler uses the CoreRT pieces in a multi-threaded fashion to validate a lot of it.

Reply:
How about a thread-static we store in the ready-to-run specific piece of CorInfoImpl that gets set when we start to compile a method. It can be passed in as an argument to the NodeFactory methods which need to do token lookups as a module / mdToken tuple.

Reply:
Thread statics for things like this are always a last resort. Statics in keep around a bunch of state with an undefined lifetime. If we want to be able to support e.g. a compilation server in the future, statics will be a problem (need to do hacks to reset them, and we can't forget to do so). So far we've been avoiding them.

Reply:
In my original comment I mentioned per-thread helper objects but that's not the only possibility, one other simple possibility is that the helper objects will be per module (something like NodeFactoryModuleContext) and these will be passed from the methods being compiled through JIT to CorInfoImpl which will then use it as the helper to call. It will have the proper context for ref resolution and at some point it will call to some thread-safe methods in R2R node factory that will execute the actual (potentially multithreaded) additions to the various node maps.

Reply:
Note that R2R supports limited form of cross-module inlining today for non-versionable methods. It means that you can get tokens from multiple modules while compiling single method, and the context to resolve the token has to always travel along with the token itself.

Reply:
I'll be happy to start looking into it once I get back from vacation. I'll probably need to better understand what exactly happens in JIT when this form of inlining occurs. As you yourself said before, we cannot generally conjure external tokens out of thin air. Arguably we may be fine to do so in case of non-versionable (cast in stone) types but today we don't have means for doing this one way or another as it would require editing the metadata. So what happens when a non-versionable method from Corelib gets inlined into a user method and in the middle it asks NodeFactory to create the MethodCodeNode for a Corelib-local methoddef e.g. Console.WriteLine (this is probably a stupid example but anyway). How are we going to represent this external call within the main app that doesn't include any methodref to Console.WriteLine?

Reply:
> middle it asks NodeFactory to create the MethodCodeNode for a Corelib-local methoddef 

In CoreCLR/crossgen today, this will cause the whole inlining attempt to be aborted.

Reply:
OK, so if I may sum up my understanding of this issue so far, once we manage to wire up module-token pairs properly and once we can edit the metadata so that we can augment the tables by adding a new synthetic TypeRef / MethodRef for non-versionable entities in available reference assemblies, we may be able to exceed Crossgen R2R inlining quality.